### PR TITLE
shrink can work on 'ancient' slots

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
If ancient packing mixes together accounts from old (cold) and new (hotter) slots, we end up rewriting accounts much more often than we'd like to.

#### Summary of Changes
Allow shrink to work on ancient storages. If we can make them smaller, then accounts of the same hot/coldness/age are kept together. This is especially true of storages that have already been packed. This change will be in conjunction with other tweaks to the tuning of how we treat ancient storages. The goal is to keep cold stuff together so more of the system remains untouched for longer.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -2038,6 +2040,7 @@ pub struct ShrinkStats {
     accounts_loaded: AtomicU64,
     purged_zero_lamports: AtomicU64,
     accounts_not_found_in_index: AtomicU64,
+    shrinking_ancient: AtomicU64,
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Since we have `num_slots_shrunk`, can this be named similarly? How about `num_ancient_slots_shrunk`?

**Review by HaoranYi — DISMISSED** [review]
lgtm

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -5086,6 +5081,15 @@ impl AccountsDb {
             shrink_slots
                 .into_par_iter()
                 .for_each(|(slot, slot_shrink_candidate)| {
+                    if self
+                        .ancient_append_vec_offset
+                        .map(|_| slot < oldest_non_ancient_slot)
+                        .unwrap_or_default()
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Why do a `map` on `ancient_append_vec_offset` if it is not used?

```suggestion
                    if slot < oldest_non_ancient_slot
```

File: accounts-db/src/accounts_db.rs [hunk 3]
```diff
@@ -5086,6 +5081,15 @@ impl AccountsDb {
...
+                    {
+                        self.shrink_stats
+                            .shrinking_ancient
+                            .fetch_add(1, Ordering::Relaxed);
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Could this atomic be moved outside the parallel loop so that we only do it once?

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: first, this is rare to hit. Second, we'd need an atomic or something to move it outside a parallel loop, right?

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: i reworked the logic. I only want to log if we have an ancient append vec offset.

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: done

**Review by HaoranYi — APPROVED** [review]
lgtm.

File: accounts-db/src/accounts_db.rs (hunk 3 above)
Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Ok, if it is rare that helps.

To move out of the parallel loop, I'd do a map/reduce, so each thread would have a local counter, then sum in the reduce step at the end, then do a single atomic update.

**Review by brooksprumo — APPROVED** [review]
:shipit:

**Pull request merged**